### ISSUE TITLE
[BUG] SKTouchAction.Cancelled not issued on iOS when SkCanvasView is in a ScrollView

### ISSUE THREAD

**Issue by MaurizioPanzica** [mono/SkiaSharp#1048]
**Description**

[On iOS]
When a SkCanvasView is in a ScrollView, and the EnableTouchEvents property is set to 'true', the touch event is not cancelled (i.e.: SkTouchAction.Cancelled is not invoked) if the ScrollView is scrolled while touching the SkCanvasView. Neither is the SkTouchAction.Released event.
Everything works fine on Android.

**Code**

See attached sample project.

**Expected Behavior**

SKTouchAction.Cancelled event invoked when the container is scrolled while touching the SkCanvasView.

**Actual Behavior**

Event is not invoked on iOS. Works fine on Android. Not tested on other platforms.

**Basic Information**

- Version with issue:  1.68.1
- Last known good version:  unknown
- IDE:  VS2019 Mac/Windows
- Platform Target Frameworks: 
  - iOS:  

**Reproduction Link**

[TestSkiaSharp.zip](https://github.com/mono/SkiaSharp/files/3916183/TestSkiaSharp.zip)

**Comment by MaurizioPanzica** [comment]
Well, after further investigations this behavior seems to be related to iOS and not to SkiaSharp itself.
In this particular scenario, the underlying scrollview, as soon as the scroll begins, invokes the method 'touchesCancelled' of the view and not the GestureRecognizer's.
This is why the SKCanvasView does not receive it. But the native view does. I guess something can be made at that level.

**Comment by daltonks** [comment]
Hey @MaurizioPanzica,
I am currently running into this problem. I need a ListView that contains SKCanvasViews with touch handling.
Have you been able to find a workaround?

**Comment by mpanzica** [comment]
@daltonks Unfortunately not.
Since then the project has been put on hold because of Covid and I've been working on another one. I guess it will be resumed in the near future. Rest assured that I will post the workaround, should I find one.

Cheers,
M.

**Comment by daltonks** [comment]
@mpanzica Thanks for the reply :)
I'm kind of led to believe that Xamarin is mucking up the touches somehow, because as far as I can tell, TouchesCancelled should be propagated to gesture recognizers when a scroll view sends the event.

**Comment by mpanzica** [comment]
Nope, if I recall well Xamarin was not guilty at all. The "cancelled" event never reaches the gesture recognizer handler... hence it is not propagated any further. I will be able to resume the project as soon as I'll be allowerd to go back to office. I can't access the client's repository from home. :-)

**Comment by daltonks** [comment]
Thanks for the info @mpanzica ! I am currently thinking that this might be because `SKTouchHandler` doesn't change its `UIGestureRecognizer.State`.

https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/implementing_a_custom_gesture_recognizer
https://github.com/mono/SkiaSharp/blob/master/source/SkiaSharp.Views.Forms/SkiaSharp.Views.Forms.iOS/SKTouchHandler.cs

**Comment by daltonks** [comment]
Nope, that idea led nowhere 🙃. Setting the `State` stops the scroll view from receiving the touches.

**Comment by daltonks** [comment]
@mpanzica I found a way that seems to work! If you keep track of the touches with `HashSet<UITouch> _touches`, you can manually cancel the ones in the gesture recognizer's `Reset` method.

https://gist.github.com/daltonks/f01cf125b580ca8b7b9213413f9299a5

**Comment by mpanzica** [comment]
Thanks for sharing, Dalton! It's night over here... I will give your
solution a try tomorrow morning before anything else! :-)

Il lun 29 giu 2020, 22:24 Dalton Spillman <notifications@github.com> ha
scritto:

**Comment by daltonks** [comment]
Cool! I made a repo based off of your reproduction .zip: https://github.com/daltonks/FixedSKTouchHandler

Tabs/spaces got all screwy, and I did strange things to use my own `SKTouchHandler`, but you can see the majority of the changes here: https://github.com/daltonks/FixedSKTouchHandler/blob/master/TestSkiaSharp.iOS/Renderers/TestSKCanvasViewRenderer.cs

The key part is keeping track of the touches and overriding `UIGestureRecognizer.Reset`.

**Comment by mattleibow** [comment]
Thanks for all the back and forth on this. I'll try to get this into the next release.

Has anyone been using this? Do all the touches that are started get the opposite finish/cancel event? Is there a potential for a memory leak or persistent touch objects?

**Comment by daltonks** [comment]
I've been using this successfully in production for a bit over 3 months now. I haven't seen any touch issues, but it would be good to look at how it behaves in a ScrollView as well as when iOS gestures take over (like minimizing the app, and maybe some other ones).

**Comment by mattleibow** [comment]
OK, thanks for the comment. I'll try a few things.

**Comment by DuncWatts** [comment]
I fixed this issue a while ago on a UIGestureRecognizer by returning false to the ShouldRecognizeSimultaneously delegate, and also triggering the Cancelled at the same time.  

I'm trying to switch over to use the SkiaSharp OnTouch and now hitting this issue. This may be a simpler fix than tracking the touches.

**Comment by daltonks** [comment]
@DuncWatts Would returning false for `ShouldRecognizeSimultaneously` completely stop scrolling if the canvas was in a scroll view and you pressed on the SkiaSharp view? That is not what I typically need, but I could see a use case for that.

**Comment by DuncWatts** [comment]
No, it allows the scroll gesture on the scroll view to take priority and cancels the touch event on the canvas, which is the functionality I was after.

**Comment by daltonks** [comment]
@DuncWatts I tried `ShouldRecognizeSimultaneously = (recognizer, gestureRecognizer) => false;` in my example project https://github.com/daltonks/FixedSKTouchHandler, but it didn't call Cancelled on scrolling. I could be doing it wrong though.

https://github.com/daltonks/FixedSKTouchHandler/blob/526c035459193d1d4d600e500653446ac510f646/TestSkiaSharp.iOS/Renderers/TestSKCanvasViewRenderer.cs#L47
